File: scratch/radical_inverse.cpp
// Online C++ compiler to run C++ program online
#include <iostream>

static constexpr int PrimeTableSize = 17;
int Primes[PrimeTableSize] = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59};
float OneMinusEpsilon = float(0x1.fffffep-1);
float RadicalInverse(int baseIndex, uint64_t a) {
    unsigned int base = Primes[baseIndex];
    // We have to stop once reversedDigits is >= limit since otherwise the
    // next digit of |a| may cause reversedDigits to overflow.
    uint64_t limit = ~0ull / base - base;
    float invBase = (float)1 / (float)base, invBaseM = 1;
    uint64_t reversedDigits = 0;
    while (a && reversedDigits < limit) {
        // Extract least significant digit from _a_ and update _reversedDigits_
        uint64_t next = a / base;
        uint64_t digit = a - next * base;
        reversedDigits = reversedDigits * base + digit;
        invBaseM *= invBase;
        a = next;
    }
    return std::min(reversedDigits * invBaseM, OneMinusEpsilon);
}

int main() {
    // Write C++ code here
    std::cout <<  RadicalInverse(0,500);

    return 0;
}

File: scratch/low_disCRAPency/radical_inverse.jl
const Primes::Vector{Int64} = Int64[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59]

function radical_inverse(prime_index::Int64, a::UInt64)::Float64
    base = Primes[prime_index]
    limit::UInt64 = typemax(UInt64) / base - base
    inv_base::Float64 = 1.0 / base
    inv_base_M::Float64 = 1.0
    reversed_digits::UInt64 = UInt64(0)
    while ((a >= 1) && (reversed_digits < limit))
        next::UInt64 = a ÷ base
        digit::UInt64 = a - next * base
        reversed_digits = reversed_digits * base + digit
        inv_base_M *= inv_base
        a = next
    end
    return min(reversed_digits * inv_base_M, 1.0-eps());
end

print(radical_inverse(1, UInt64(3)))